File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/SymbolVisibilityFilterController.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import com.esri.arcgisruntime.mapping.view.MapScaleChangedEvent;
import com.esri.arcgisruntime.mapping.view.MapScaleChangedListener;
import com.esri.arcgisruntime.mapping.view.MapView;

import java.util.HashMap;
import java.util.Map;

public class SymbolVisibilityFilterController implements MapScaleChangedListener {
    private final Map<GraphicControllerSink, SymbolVisibilityFilter> graphicControllers = new HashMap<>();
    private final Map<GraphicControllerSink, Boolean> initialValues = new HashMap<>();
    private final FlutterMapViewDelegate flutterMapViewDelegate;

    private boolean isRegister;

    public SymbolVisibilityFilterController(FlutterMapViewDelegate flutterMapViewDelegate) {
        this.flutterMapViewDelegate = flutterMapViewDelegate;
    }

    public void clear() {
        flutterMapViewDelegate.removeMapScaleChangedListener(this);
        isRegister = false;

        for (Map.Entry<GraphicControllerSink, SymbolVisibilityFilter> entry :
                graphicControllers.entrySet()) {
            entry.getKey().setVisible(initialValues.remove(entry.getKey()));
        }
        graphicControllers.clear();
    }


    public boolean containsGraphicsController(GraphicControllerSink graphicController) {
        return graphicControllers.containsKey(graphicController);
    }

    public void updateInitialVisibility(GraphicControllerSink graphicController, Boolean initValue) {
        if (initialValues.containsKey(graphicController)) {
            initialValues.replace(graphicController, initValue);
        }
    }

    public void invalidate(GraphicControllerSink graphicController) {
        if (containsGraphicsController(graphicController)) {
            handleGraphicsFilterZoom(graphicController, graphicControllers.get(graphicController), flutterMapViewDelegate.getMapScale());
        }
    }

    public void addGraphicsController(GraphicControllerSink graphicController, SymbolVisibilityFilter symbolVisibilityFilter, boolean initValue) {
        if (initialValues.containsKey(graphicController)) {
            initialValues.remove(graphicController);
        }

        initialValues.put(graphicController, initValue);

        handleGraphicsFilterZoom(graphicController, symbolVisibilityFilter, flutterMapViewDelegate.getMapScale());

        if (graphicControllers.containsKey(graphicController)) {
            if (graphicControllers.get(graphicController) == symbolVisibilityFilter) {
                return;
            }
            graphicControllers.remove(graphicController);
        }

        graphicControllers.put(graphicController, symbolVisibilityFilter);
        handleRegistrationToScaleChanged();
    }

    public void removeGraphicsController(GraphicControllerSink graphicController) {
        if (!containsGraphicsController(graphicController)) {
            return;
        }
        graphicController.setVisible(initialValues.remove(graphicController));
        graphicControllers.remove(graphicController);
        handleRegistrationToScaleChanged();
    }

    @Override
    public void mapScaleChanged(MapScaleChangedEvent mapScaleChangedEvent) {
        final double currentZoom = mapScaleChangedEvent.getSource().getMapScale();
        for (Map.Entry<GraphicControllerSink, SymbolVisibilityFilter> entry :
                graphicControllers.entrySet()) {
            handleGraphicsFilterZoom(entry.getKey(), entry.getValue(), currentZoom);
        }
    }


    private void handleGraphicsFilterZoom(GraphicControllerSink graphicController, SymbolVisibilityFilter visibilityFilter, double currentZoom) {
        if (Double.isNaN(currentZoom)) {
            return;
        }
        if (currentZoom < visibilityFilter.getMinZoom() && currentZoom > visibilityFilter.getMaxZoom()) {
            graphicController.setVisible(initialValues.get(graphicController));
        } else {
            if (!graphicController.isSelected()) {
                graphicController.setVisible(false);
            }
        }
    }

    private void handleRegistrationToScaleChanged() {
        if (graphicControllers.size() > 0 && !isRegister) {
            isRegister = true;
            flutterMapViewDelegate.addMapScaleChangedListener(this);
        } else if (graphicControllers.size() == 0 && isRegister) {
            isRegister = false;
            flutterMapViewDelegate.removeMapScaleChangedListener(this);
        }
    }
}

File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/map/SymbolVisibilityFilterController.kt
package com.valentingrigorean.arcgis_maps_flutter.map

import com.arcgismaps.mapping.view.MapView
import com.valentingrigorean.arcgis_maps_flutter.mapping.symbology.GraphicControllerSink
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

class SymbolVisibilityFilterController(
    private val mapView: MapView,
    scope: CoroutineScope
) {
    private val graphicControllers: MutableMap<GraphicControllerSink, SymbolVisibilityFilter> =
        HashMap()
    private val initialValues: MutableMap<GraphicControllerSink, Boolean> = HashMap()

    init {
        mapView.mapScale.onEach {
            invalidate()
        }.launchIn(scope)
    }

    fun clear() {
        graphicControllers.clear()
    }

    fun containsGraphicsController(graphicController: GraphicControllerSink): Boolean {
        return graphicControllers.containsKey(graphicController)
    }

    fun updateInitialVisibility(graphicController: GraphicControllerSink, initValue: Boolean) {
        if (initialValues.containsKey(graphicController)) {
            initialValues.replace(graphicController, initValue)
        }
    }

    fun invalidate(graphicController: GraphicControllerSink) {
        if (containsGraphicsController(graphicController)) {
            handleGraphicsFilterZoom(
                graphicController,
                graphicControllers[graphicController]!!,
                mapView.mapScale.value
            )
        }
    }

    fun addGraphicsController(
        graphicController: GraphicControllerSink,
        symbolVisibilityFilter: SymbolVisibilityFilter,
        initValue: Boolean
    ) {
        if (initialValues.containsKey(graphicController)) {
            initialValues.remove(graphicController)
        }
        initialValues[graphicController] = initValue
        handleGraphicsFilterZoom(
            graphicController,
            symbolVisibilityFilter,
            mapView.mapScale.value
        )
        if (graphicControllers.containsKey(graphicController)) {
            if (graphicControllers[graphicController] === symbolVisibilityFilter) {
                return
            }
            graphicControllers.remove(graphicController)
        }
        graphicControllers[graphicController] = symbolVisibilityFilter
    }

    fun removeGraphicsController(graphicController: GraphicControllerSink) {
        if (!containsGraphicsController(graphicController)) {
            return
        }
        graphicController.visible = initialValues.remove(graphicController) ?: true
        graphicControllers.remove(graphicController)
    }


    fun invalidate() {
        for ((key, value) in graphicControllers) {
            handleGraphicsFilterZoom(key, value, mapView.mapScale.value)
        }
    }

    private fun handleGraphicsFilterZoom(
        graphicController: GraphicControllerSink,
        visibilityFilter: SymbolVisibilityFilter,
        currentZoom: Double
    ) {
        if (java.lang.Double.isNaN(currentZoom)) {
            graphicController.visible = initialValues[graphicController] ?: true
            return
        }
        if (currentZoom < visibilityFilter.minZoom || currentZoom > visibilityFilter.maxZoom) {
            graphicController.visible = false
        } else {
            // visible when it's in the zoom range or if it's selected
            graphicController.visible = graphicController.isSelected || initialValues[graphicController] ?: true
        }
    }

}